build/deps/versions.bzl

```diff
@@ -0,0 +1,182 @@
+    if ncrypto_start == -1:
+        fail("Could not find ncrypto strip_prefix in WORKSPACE")
+
+    ncrypto_start += len('strip_prefix = "ncrypto-')
+    ncrypto_end = workspace_content.find('"', ncrypto_start)
+    if ncrypto_end == -1:
+        fail("Could not parse ncrypto version in WORKSPACE")
+
+    ncrypto_version = workspace_content[ncrypto_start:ncrypto_end]
+
+    # Read V8 version
+    v8_bzl_content = ctx.read(ctx.path(ctx.workspace_root.get_child("build").get_child("deps").get_child("v8.bzl")))
+
+    v8_start = v8_bzl_content.find('VERSION = "')
+    if v8_start == -1:
+        fail("Could not find VERSION in v8.bzl")
+
+    v8_start += len('VERSION = "')
+    v8_end = v8_bzl_content.find('"', v8_start)
+    if v8_end == -1:
+        fail("Could not parse VERSION in v8.bzl")
+
+    v8_version = v8_bzl_content[v8_start:v8_end]
+
+    # Generate the versions.h header file
+    header_content = """// This file is auto-generated. Do not edit manually.
+#pragma once
+
+#include <kj/string.h>
+
+namespace workerd::api::node {{
+
+// Version of the ada-url library
+static constexpr kj::StringPtr adaVersion = "{ada_version}"_kj;
+
+// Version of Node.js
+static constexpr kj::StringPtr nodeVersion = "{node_version}"_kj;
+
+// Version of Brotli compression library
+static constexpr kj::StringPtr brotliVersion = "{brotli_version}"_kj;
+
+// Version of simdutf library
+static constexpr kj::StringPtr simdutfVersion = "{simdutf_version}"_kj;
+
+// Version of SQLite database
+static constexpr kj::StringPtr sqliteVersion = "{sqlite_version}"_kj;
+
+// Version of nbytes library
+static constexpr kj::StringPtr nbytesVersion = "{nbytes_version}"_kj;
+
+// Version of ncrypto library
+static constexpr kj::StringPtr ncryptoVersion = "{ncrypto_version}"_kj;
+
+// Version of V8 JavaScript engine
+static constexpr kj::StringPtr v8Version = "{v8_version}"_kj;
+
+}}  // namespace workerd::api::node
+""".format(ada_version = ada_version, node_version = node_version, brotli_version = brotli_version, simdutf_version = simdutf_version, sqlite_version = sqlite_version, nbytes_version = nbytes_version, ncrypto_version = ncrypto_version, v8_version = v8_version)
+
+    # Write the header file
+    ctx.file("versions.h", header_content)
+
+    # Create a BUILD file to export the header
+    build_content = """
+exports_files(["versions.h"], visibility = ["//visibility:public"])
+
+cc_library(
+    name = "versions",
+    hdrs = ["versions.h"],
+    visibility = ["//visibility:public"],
+)
+"""
+
+    ctx.file("BUILD.bazel", build_content)
+
+_extract_versions = repository_rule(
+    implementation = _extract_versions_impl,
+    doc = "Extracts versions from dependency bzl files and generates C++ headers",
+    attrs = {
+        "node_version": attr.string(
+            mandatory = True,
+            doc = "The Node.js version string",
+        ),
+    },
+)
+
+def extract_versions(node_version):
```

Comment:
Question: how is the node_version extracted? 
Is the version of Node in this repo?
If yes, could you expand on how that's related to workerd?

FYI in unenv we set the version to whatever API unenv implements. I honestly have no idea what would be the best value for workerd to use?

Reply:
I shared the NODE_VERSION we define in our main workspace file - https://github.com/cloudflare/workerd/blob/main/WORKSPACE#L21.

It does give a very new meaning to the term though, but seems a relatively good proxy for "the current node version".

We may wish to separate these definitions or more clearly define some process here for keeping this updated. Alternatively we may just always want to reflect the latest LTS instead and have a Bazel step that looks that up and uses that.

Fetching the latest LTS at build time might be the better option here? I can look into that.

Reply:
Hmm..the version of Node.js we use for the build is not the same as the version of the Node.js we likely want to claim interop for via the API. I don't think we should be using the build as the source here. We need to think through the version reporting a bit more.

Reply:
I see two options here:

1. We maintain a Node version header file, tracking a baseline release to which we support its API. We then need to keep this file manually updated.
2. We add a build step to fetch the latest Node.js LTS version and setup the header automatically.

I'm open to either, this is about how we want to maintain it going forward.

Reply:
I think option 1 makes the most sense at this point.